clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -86,6 +85,25 @@ public AbortedTransaction(long pid, long firstOffset) {
         }
 
         @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass())
+                return false;
+
+            AbortedTransaction that = (AbortedTransaction) o;
+
+            return pid == that.pid && firstOffset == that.firstOffset;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = (int) (pid ^ (pid >>> 32));
```

Comment:
I think this is probably not the proper forum to debate the merits of an idiomatic java pattern that is widely used in the codebase. If you feel strongly, open a JIRA and we can have a wider discussion there. In this case, the addition of `equals` and `hashCode` just facilitates testing. We would implement them in other request types if and when we need them (my guess is we will start to need them more as we have now moved `Struct` outside of the classes).